File: src/main/java/brs/services/AccountService.java
package brs.services;

import brs.Account;
import brs.Account.AccountAsset;
import brs.Account.Event;
import brs.Account.RewardRecipientAssignment;
import brs.AssetTransfer;
import brs.util.Observable;

import java.util.Collection;
import java.util.function.Consumer;

public interface AccountService extends Observable<Account, Event> {
  boolean addAssetListener(Consumer<AccountAsset> listener, Event eventType);

  Account getAccount(long id);

  Account getAccount(long id, int height);

  Account getAccount(byte[] publicKey);

  Collection<AssetTransfer> getAssetTransfers(long accountId, int from, int to);

  Collection<AccountAsset> getAssets(long accountId, int from, int to);

  Collection<RewardRecipientAssignment> getAccountsWithRewardRecipient(Long recipientId);

  Collection<Account> getAccountsWithName(String name);

  Collection<Account> getAllAccounts(int from, int to);

  Account getOrAddAccount(long id);

  void flushAccountTable();

  int getCount();

  void addToForgedBalanceNQT(Account account, long amountNQT);

  void setAccountInfo(Account account, String name, String description);

  void addToAssetBalanceQNT(Account account, long assetId, long quantityQNT);

  void addToUnconfirmedAssetBalanceQNT(Account account, long assetId, long quantityQNT);

  void addToAssetAndUnconfirmedAssetBalanceQNT(Account account, long assetId, long quantityQNT);

  void addToBalanceNQT(Account account, long amountNQT);

  void addToUnconfirmedBalanceNQT(Account account, long amountNQT);

  void addToBalanceAndUnconfirmedBalanceNQT(Account account, long amountNQT);

  RewardRecipientAssignment getRewardRecipientAssignment(Account account);

  void setRewardRecipientAssignment(Account account, Long recipient);

  long getUnconfirmedAssetBalanceQNT(Account account, long assetId);
}

File: src/main/java/brs/services/AccountService.kt
package brs.services

import brs.Account
import brs.Account.*
import brs.AssetTransfer
import brs.util.Observable

interface AccountService : Observable<Account, Event> {
    suspend fun getCount(): Int

    suspend fun addAssetListener(eventType: Event, listener: suspend (AccountAsset) -> Unit)

    suspend fun getAccount(id: Long): Account?

    suspend fun getAccount(id: Long, height: Int): Account?

    suspend fun getAccount(publicKey: ByteArray): Account?

    suspend fun getAssetTransfers(accountId: Long, from: Int, to: Int): Collection<AssetTransfer>

    suspend fun getAssets(accountId: Long, from: Int, to: Int): Collection<AccountAsset>

    suspend fun getAccountsWithRewardRecipient(recipientId: Long?): Collection<RewardRecipientAssignment>

    suspend fun getAccountsWithName(name: String): Collection<Account>

    suspend fun getAllAccounts(from: Int, to: Int): Collection<Account>

    suspend fun getOrAddAccount(id: Long): Account

    suspend fun flushAccountTable()

    // TODO rename methods
    suspend fun addToForgedBalanceNQT(account: Account, amountNQT: Long)

    suspend fun setAccountInfo(account: Account, name: String, description: String)

    suspend fun addToAssetBalanceQNT(account: Account, assetId: Long, quantityQNT: Long)

    suspend fun addToUnconfirmedAssetBalanceQNT(account: Account, assetId: Long, quantityQNT: Long)

    suspend fun addToAssetAndUnconfirmedAssetBalanceQNT(account: Account, assetId: Long, quantityQNT: Long)

    suspend fun addToBalanceNQT(account: Account, amountNQT: Long)

    suspend fun addToUnconfirmedBalanceNQT(account: Account, amountNQT: Long)

    suspend fun addToBalanceAndUnconfirmedBalanceNQT(account: Account, amountNQT: Long)

    suspend fun getRewardRecipientAssignment(account: Account): RewardRecipientAssignment?

    suspend fun setRewardRecipientAssignment(account: Account, recipient: Long)

    suspend fun getUnconfirmedAssetBalanceQNT(account: Account, assetId: Long): Long
}
